src/lib.rs

```diff
@@ -307,6 +314,88 @@ impl Default for Options {
     }
 }
 
+#[derive(Debug)]
+/// A raw image definition which can be used to create an optimized png
+pub struct RawImage {
+    png: Arc<PngImage>,
+}
+
+impl RawImage {
+    /// Construct a new raw image definition
+    ///
+    /// * `width` - The width of the image in pixels
+    /// * `height` - The height of the image in pixels
+    /// * `color_type` - The color type of the image
+    /// * `bit_depth` - The bit depth of the image
+    /// * `data` - The raw pixel data of the image
+    pub fn new(
+        width: u32,
+        height: u32,
+        color_type: ColorType,
+        bit_depth: BitDepth,
+        data: Vec<u8>,
+    ) -> Result<Self, PngError> {
+        // Validate bit depth
+        let valid_depth = match color_type {
+            ColorType::Grayscale { .. } => true,
+            ColorType::Indexed { .. } => (bit_depth as u8) <= 8,
+            _ => (bit_depth as u8) >= 8,
+        };
+        if !valid_depth {
+            return Err(PngError::InvalidDepthForType(bit_depth, color_type));
+        }
+
+        // Validate data length
+        let bpp = bit_depth as usize * color_type.channels_per_pixel() as usize;
+        let row_bytes = (bpp * width as usize + 7) / 8;
+        let expected_len = row_bytes * height as usize;
+        if data.len() != expected_len {
+            return Err(PngError::IncorrectDataLength(data.len(), expected_len));
+        }
+
+        Ok(Self {
+            png: Arc::new(PngImage {
+                ihdr: IhdrData {
+                    width,
+                    height,
+                    color_type,
+                    bit_depth,
+                    interlaced: Interlacing::None,
+                },
+                data,
+                aux_headers: IndexMap::new(),
+            }),
+        })
+    }
+
+    /// Add a png header chunk, such as "iTXt", to be included in the output
+    pub fn add_png_header(&mut self, chunk_type: [u8; 4], data: Vec<u8>) {
+        // We can guarantee this will succeed - failure indicates a bug
+        let png = Arc::get_mut(&mut self.png).unwrap();
+        png.aux_headers.insert(chunk_type, data);
+    }
+
+    /// Add an ICC profile for the image
+    pub fn add_icc_profile(&mut self, data: Vec<u8>) {
+        // Compress with default compression level
+        if let Ok(mut compressed) = deflate::deflate(&data, 11, &AtomicMin::new(None)) {
```

Comment:
Maybe take it gzipped, so that compression can be optimized too?

Reply:
Not quite sure what you're meaning, is it common for icc profiles to be provided in gzip form?
There's an existing issue #432 for recompression of iCCP chunks, if that's relevant.

Reply:
Yes, the motivation was to be able to optimize compression of iCCP. I suppose this method is not necessary for it, and an optimizer can use `add_png_chunk(*b"iCCP` instead.


BTW: if this method is not keeping `data` value permanently, then it shouldn't be taking ownership via `Vec`, and take a `&[u8]` slice instead.

Reply:
Yeah if you have an existing iCCP chunk you should use `add_png_chunk`. Future work can implement recompression, independently of this api.

Ah, good call on the slice. I've changed this now.